marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Database/PostgreSQL.hs

```diff
@@ -0,0 +1,564 @@
+    decodeResults [] = Right Nothing
+    decodeResults rows@((hash, _, _, _) : _) =
+      Just . GenesisBlock (HeaderHash $ toShort hash) . Set.fromList <$> traverse decodeGenesisTx rows
+
+    decodeGenesisTx :: (ByteString, ByteString, Int64, ByteString) -> Either Text GenesisTx
+    decodeGenesisTx (_, txId, lovelace, address) = GenesisTx
+      <$> decodeTxId txId
+      <*> pure (fromIntegral lovelace)
+      <*> decodeAddressAny address
+
+-- GetHeaderAtPoint
+
+getHeaderAtPoint :: GetHeaderAtPoint Session
+getHeaderAtPoint = GetHeaderAtPoint \case
+  ChainPointAtGenesis -> pure Origin
+  point@(ChainPoint slotNo hash) ->
+    statement (slotNoToParam slotNo, headerHashToParam hash) $ refineResult decodeResults
+      [maybeStatement|
+        SELECT block.blockNo :: bigint
+          FROM chain.block AS block
+        WHERE block.slotNo = $1 :: bigint
+          AND block.id     = $2 :: bytea
+      |]
+    where
+      decodeResults :: Maybe Int64 -> Either Text (WithOrigin BlockHeader)
+      decodeResults Nothing        = Left $ "No block found at " <> T.pack (show point)
+      decodeResults (Just blockNo) = Right $ At $ BlockHeader slotNo hash $ BlockNo $ fromIntegral blockNo
+
+decodeTxId :: ByteString -> Either Text TxId
+decodeTxId txId = case deserialiseFromRawBytes AsTxId txId of
+  Nothing    -> Left "Invalid TxId"
+  Just txId' -> Right txId'
+
+decodeAddressAny :: ByteString -> Either Text AddressAny
+decodeAddressAny address = case deserialiseFromRawBytes AsAddressAny address of
+  Nothing       -> Left "Invalid address"
+  Just address' -> Right address'
+
+-- GetIntersectionPoints
+
+getIntersectionPoints :: GetIntersectionPoints Session
+getIntersectionPoints = GetIntersectionPoints $ statement () $ rmap decodeResults
+  [vectorStatement|
+    SELECT slotNo :: bigint, id :: bytea
+    FROM   chain.block
+    WHERE  rollbackToBlock IS NULL
+    ORDER BY slotNo DESC LIMIT 2160
+  |]
+  where
+    decodeResults :: Vector (Int64, ByteString) -> [ChainPoint]
+    decodeResults = fmap decodeResult . V.toList
+
+    decodeResult :: (Int64, ByteString) -> ChainPoint
+    decodeResult (slotNo, hash) = ChainPoint (SlotNo $ fromIntegral slotNo) (HeaderHash $ toShort hash)
+
+-- CommitRollback
+
+commitRollback :: CommitRollback Session
+commitRollback = CommitRollback \case
+  ChainPointAtGenesis -> sql $ BS.intercalate "\n"
+    [ "BEGIN;"
+    , "TRUNCATE TABLE chain.tx;"
+    , "TRUNCATE TABLE chain.txOut;"
+    , "TRUNCATE TABLE chain.txIn;"
+    , "DELETE FROM chain.asset;"
+    , "TRUNCATE TABLE chain.assetOut;"
+    , "TRUNCATE TABLE chain.assetMint;"
+    , "TRUNCATE TABLE chain.block;"
+    , "COMMIT;"
+    ]
+  ChainPoint slotNo hash -> statement (slotNoToParam slotNo, headerHashToParam hash)
+    [resultlessStatement|
+      WITH blockUpdates AS
+        ( UPDATE chain.block
+             SET rollbackToSlot  = $1 :: bigint
+               , rollbackToBlock = $2 :: bytea
+           WHERE slotNo > $1 :: bigint
```

Comment:
Should this be `>=` or `>`? My memory is that the rollback messages report a new block that should replace the old one at the slot and where everything subsequent should be discarded. Does the rollback strategy replace the block at the rollback slot and delete the transactions there? Maybe I'm mistaken.

Reply:
I think you're mistaken - I'm pretty confident that the rollback `point` indicates the most recent block that is _not_ rolled back.

e.g. If the client has blocks `A B C D E F` and receives a `MsgRollBackward C`, then the client should delete everything after `C`, leaving `A B C`. The next `MsgRollForward` will return the next block `D'` after `C` in the new fork of the chain.

Reply:
My evidence for this is that the chain sync server always starts by sending a `MsgRollback` to the client's starting block You will start by receiving a `MsgRollback ChainPointAtGenesis` if you do not start from an intersection point. If you receive a `MsgIntersectFound p` for some point `p` then send a `MsgRequestNext`, you will always receive a `MsgRollBackward p tip` before you receive a `MsgRollForward`.

Also, the `MsgRollBackward` only sends the `point`, not the block, so it doesn't tell you to "replace the old one at this point."